anvil-python/anvil/commands/upgrades/intra_channel.py

```diff
@@ -0,0 +1,190 @@
+        """
+        deployed_machine_apps = self.get_charm_deployed_versions("controller")
+
+        all_deployed_apps = deployed_machine_apps.copy()
+        LOG.debug(f"All deployed apps: {all_deployed_apps}")
+        if self.is_track_changed_for_any_charm(all_deployed_apps):
+            error_msg = "MAAS-Anvil cannot upgrade across tracks! Please modify refresh manifest."
+            return Result(ResultType.FAILED, error_msg)
+
+        self.refresh_apps(deployed_machine_apps, "controller")
+        return Result(ResultType.COMPLETED)
+
+
+class LatestInChannelCoordinator:
+    """Coordinator for refreshing charms in their current channel."""
+
+    def __init__(
+        self,
+        deployment: Deployment,
+        client: Client,
+        jhelper: JujuHelper,
+        manifest: Manifest,
+    ):
+        """Upgrade coordinator.
+
+        Execute plan for conducting an upgrade.
+
+        :client: Helper for interacting with clusterd
+        :jhelper: Helper for interacting with pylibjuju
+        :manifest: Manifest object
+        """
+        self.deployment = deployment
+        self.client = client
+        self.jhelper = jhelper
+        self.manifest = manifest
+        self.preseed = self.manifest.deployment_config
+
+    def run_plan(self) -> None:
+        """Execute the upgrade plan."""
+        plan = self.get_plan()
+        run_plan(plan, console)
+
+    def get_plan(self) -> list[BaseStep]:
+        """Return the upgrade plan."""
+        plan: list[BaseStep] = []
+        plan.append(LatestInChannel(self.jhelper, self.manifest))
+
+        plan.extend(
+            haproxy_upgrade_steps(
+                self.client,
+                self.manifest,
+                self.jhelper,
+                self.deployment.infrastructure_model,
+                self.preseed,
+            )
+        )
+        plan.extend(
+            postgresql_upgrade_steps(
+                self.client,
+                self.manifest,
+                self.jhelper,
+                self.deployment.infrastructure_model,
+                self.preseed,
+            )
+        )
+        plan.extend(
+            maas_region_upgrade_steps(
+                self.client,
+                self.manifest,
+                self.jhelper,
+                self.deployment.infrastructure_model,
+            )
+        )
+        plan.extend(
+            maas_agent_upgrade_steps(
+                self.client,
+                self.manifest,
+                self.jhelper,
+                self.deployment.infrastructure_model,
+            )
+        )
+
+        # TODO: Update MAAS-Anvil sunbeam tag to allow using
+        # sunbeam.commands.upgrades.base.UpgradeFeatures instead of
+        # sunbeam.commands.upgrades.base.UpgradePlugins
+        # plan.append(
+        #     UpgradeFeatures(self.deployment, upgrade_release=False),
+        # )
+        plan.append(UpgradePlugins(self.deployment, upgrade_release=False))
```

Comment:
Unfortunately, you have to implement our own `UpgradePlugins` that is using Anvil's `PluginManager` instead of Sunbeam: `anvil-python/anvil/jobs/plugin.py:PluginManager`

Reply:
The good news is that we are talking about 20 new lines in the `upgrades/based.py` which you have to also create (with a license header)